packages/vitest-pool-workers/src/pool/index.ts

```diff
@@ -0,0 +1,767 @@
+	// Some package managers don't enforce `peerDependencies` requirements,
+	// so add a runtime sanity check to ensure things don't break in strange ways.
+	const poolPkgJson = getPackageJson(__dirname);
+	const vitestPkgJson = getPackageJson(
+		path.dirname(ctx.projectFiles.workerPath)
+	);
+	assert(
+		poolPkgJson !== undefined,
+		"Expected to find `package.json` for `@cloudflare/vitest-pool-workers`"
+	);
+	assert(
+		vitestPkgJson !== undefined,
+		"Expected to find `package.json` for `vitest`"
+	);
+
+	const expectedVitestVersion = poolPkgJson.peerDependencies?.vitest;
+	const actualVitestVersion = vitestPkgJson.version;
+	assert(
+		expectedVitestVersion !== undefined,
+		"Expected to find `@cloudflare/vitest-pool-workers`'s `vitest` version constraint"
+	);
+	assert(
+		actualVitestVersion !== undefined,
+		"Expected to find `vitest`'s version"
+	);
+
+	if (expectedVitestVersion !== actualVitestVersion) {
+		const message = [
+			`You're running \`vitest@${actualVitestVersion}\`, but this version of \`@cloudflare/vitest-pool-workers\` only supports \`vitest@${expectedVitestVersion}\`.`,
+			"`@cloudflare/vitest-pool-workers` currently depends on internal Vitest APIs that are not protected by semantic-versioning guarantees.",
+			`Please install \`vitest@${expectedVitestVersion}\` to continue using \`@cloudflare/vitest-pool-workers\`.`,
+		].join("\n");
+		throw new Error(message);
+	}
+}
+
+export default function (ctx: Vitest): ProcessPool {
+	assertCompatibleVitestVersion(ctx);
+
+	return {
+		name: "vitest-pool-workers",
+		async runTests(specs, invalidates) {
+			// Vitest waits for the previous `runTests()` to complete before calling
+			// `runTests()` again:
+			// https://github.com/vitest-dev/vitest/blob/v1.0.4/packages/vitest/src/node/core.ts#L458-L459
+			// This behaviour is required for stacked storage to work correctly.
+			// If we had concurrent runs, stack pushes/pops would interfere. We should
+			// always have an empty, fully-popped stacked at the end of a run.
+
+			// 1. Collect new specs
+			const parsedProjectOptions = new Set<WorkspaceProject>();
+			for (const [project, testFile] of specs) {
+				// Vitest validates all project names are unique
+				const projectName = project.getName();
+				let workersProject = allProjects.get(projectName);
+				// Parse project options once per project per re-run
+				if (workersProject === undefined) {
+					workersProject = {
+						project,
+						options: await parseProjectOptions(project),
+						testFiles: new Set(),
+						relativePath: getRelativeProjectPath(project.path),
+					};
+					allProjects.set(projectName, workersProject);
+				} else if (!parsedProjectOptions.has(project)) {
+					workersProject.project = project;
+					workersProject.options = await parseProjectOptions(project);
+					workersProject.relativePath = getRelativeProjectPath(project.path);
+				}
+				workersProject.testFiles.add(testFile);
+
+				parsedProjectOptions.add(project);
+			}
```

Comment:
I'm not entirely sure what this block of code is doing. From the comments, it seems like there's a cache for projects across re-runs, but I'm not sure how `parsedProjectOptions` plays into that. Is it a cache for project options per run, in case a project is specified in multiple specs? What's the purpose of a top-level cache if `project`, `options`, and `relativePath` are re-generated per run?